### DISCUSSION TITLE
Forwarding x-grafana headers when using protocol http

### DISCUSSION THREAD

**Discussion by shachakz** [grafana/clickhouse-datasource#667] [Ideas]
Hi 👋 

I'm currently working on a project that involves passing Grafana queries through an authorization proxy. This allows us to restrict the data accessible to each user. We've successfully implemented this with various data sources by enabling the global configuration flag [dataproxy.send_user_header](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#send_user_header) on, so the user id is attached as a header on every Grafana request.

**Currently, the clickhouse-datasource plugin doesn't support this out of the box, and I would like to add this to support forwarding those headers.**

I have an E2E working POC, which is implemented as such:
1. Turns on `ForwardHeaders` (in `DriverSettings`) configuration from sqlds if the flag `forwardHeaders` is `true` (in `DataSourceInstanceSettings`) inside the plugin `Settings` function
2. When calling Connect, parse message (json.RawMessage) and extract the headers from `sqlds.HeaderKey` and pass it to `clickhouse.Options.HttpHeaders`
3. In main.go, when creating the datasource, also set `ds.EnableMultipleConnections` to `true` if forward headers is on (otherwise the sqlds lib won't forward the headers)
4. Extract datasource creation logic from main.go to a new file under plugin package, so we can add some tests using the datasource class (instead of driver class)

I'll support changing this config from the settings UI as well. (didn't get to it yet)

I would love to know what do you think about the feature and the suggested implementation.
If it's cool, I'll be glad to contribute this feature 💪

**Comment by adamyeats** [comment]
Hey @shachakz thanks for your suggestion here! 🙌  @SpencerTorres Do you have any thoughts on this issue?

**Comment by SpencerTorres** [comment]
Thanks for the contribution! Sounds like it could be useful as long as it is something that can be toggled on/off by the user.

We do have a new version of the plugin that contains some changes involving headers in the clickhouse driver (#633)

I would recommend developing with that branch in mind since there will likely be significant merge conflicts.

Perhaps you could open a draft PR with your PoC and we can iterate from there?

**Reply by shachakz** [reply]
I added this as a "toggle-able" config as shown here:
![image](https://github.com/grafana/clickhouse-datasource/assets/11520905/11914a5d-83ca-4378-994f-ce9cd7800628)

@SpencerTorres as you suggested, I'll develop my change with this specific branch in mind and open a PR with my implementation to this feature!

**Comment by shachakz** [comment]
@SpencerTorres @adamyeats 

Here is my ready for review PR - [#670](https://github.com/grafana/clickhouse-datasource/pull/670)
After looking on https://github.com/grafana/clickhouse-datasource/pull/633, I think the merge conflict is going to be minor:

- We'll need to merge the HttpHeaders map (on Connect method) and the Settings in driver.go
- We'll need to match the changed lower case convention for the forwardHeaders label & description

**Comment by shachakz** [comment]
Hi @SpencerTorres , @adamyeats 

I saw that you merged http headers changes. I resolved all the conflicts and added the "forward grafana headers" option under the http configuration section and added some relevant tests.
Let me know if there's something else I can do to push this forward 😊

**Comment by SpencerTorres** [comment]
Merged in https://github.com/grafana/clickhouse-datasource/pull/670, available in v4.0.3

Closing this, thanks for the contribution!
